test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/Validation/DefaultObjectValidatorTests.cs

```diff
@@ -655,13 +656,40 @@ public void Validate_CollectionType_ListOfComplexType_Invalid()
             Assert.Equal(ValidationAttributeUtil.GetRequiredErrorMessage("Profession"), error.ErrorMessage);
         }
 
+        public static TheoryData<object, Type> ValidCollectionData
+        {
+            get
+            {
+                return new TheoryData<object, Type>()
+                {
+                    { new int[] { 1, 2, 3 }, typeof(int[]) },
+                    { new string[] { "Foo", "Bar", "Baz" }, typeof(string[]) },
+                    { new List<string> { "Foo", "Bar", "Baz" }, typeof(IList<string>)},
+                    { new HashSet<string> { "Foo", "Bar", "Baz" }, typeof(string[]) },
+                    {
+                        new List<DateTime>
+                        {
+                            DateTime.Parse("1/1/14"),
+                            DateTime.Parse("2/1/14"),
+                            DateTime.Parse("3/1/14"),
+                        },
+                        typeof(ICollection<DateTime>)
+                    },
+                    {
+                        new HashSet<Uri>
+                        {
+                            new Uri("http://example.com/1"),
+                            new Uri("http://example.com/2"),
+                            new Uri("http://example.com/3"),
+                        },
+                        typeof(HashSet<Uri>)
+                    },
+                };
+            }
+        }
+
         [Theory]
-        [InlineData(new[] { "Foo", "Bar", "Baz" }, typeof(string[]))]
-        [InlineData(new[] { 1, 2, 3 }, typeof(int[]))]
-        [InlineData(new[] { "Foo", "Bar", "Baz" }, typeof(IList<string>))]
-        [InlineData(new[] { "Foo", "Bar", "Baz" }, typeof(HashSet<string>))]
-        [InlineData(new[] { "1/1/14", "2/2/14", "3/3/14" }, typeof(ICollection<DateTime>))]
-        [InlineData(new[] { "Foo", "Bar", "Baz" }, typeof(HashSet<Uri>))]
+        [MemberData(nameof(ValidCollectionData))]
         public void Validate_IndexedCollectionTypes_Valid(object model, Type type)
```

Comment:
What was important was the `type` parameter, not the runtime type of `model`. But, sure, can use instances matching the checked `type`.